skdate/src/date.sk

```diff
@@ -0,0 +1,757 @@
+        days +
+        if (this.leapYear(year)) {
+          366
+        } else {
+          365
+        };
+      !year = year + 1;
+    };
+    days;
+  }
+
+  fun daysSince(refYear: Int = 1970): Int {
+    invariant(
+      refYear <= this.year,
+      "Expected date year greater than reference year",
+    );
+    days = this.dayBetween(refYear, this.year);
+    days + this.getDayOfYear() - 1;
+  }
+
+  fun daysUntil(refYear: Int = 1970): Int {
+    days = this.dayBetween(this.year + 1, refYear);
+    days +
+      if (this.leapYear(this.year)) {
+        366 - this.getDayOfYear()
+      } else {
+        365 - this.getDayOfYear()
+      };
+  }
+
+  //ISO 8601 Standard
+  fun getWeekDay(): Day {
+    days = if (this.year >= 1970) this.daysSince() + 3 else {
+      this.daysUntil()
+    };
+    Day::dayForCode((days % 7) + 1);
+  }
+
+  fun getWeekNumber(kind: WeekNumber = WNISO()): Int {
+    this.getWeekDate(kind).week;
+  }
+
+  fun compare(other: Date): Order {
+    if (this.year == other.year) {
+      if (this.month == other.month) {
+        this.day.compare(other.day)
+      } else {
+        this.month.compare(other.month)
+      }
+    } else {
+      this.year.compare(other.year)
+    }
+  }
+
+  private fun leapYear(year: Int): Bool {
+    year % 4 == 0 && (year % 100 != 0 || year % 400 == 0);
+  }
+
+  private fun p(year: Int): Int {
+    (year + (year / 4) - (year / 100) + (year / 400)) % 7;
+  }
+
+  private fun weeks(year: Int): Int {
+    52 +
+      if (this.p(year) == 4 || this.p(year - 1) == 3) {
+        1
+      } else {
+        0
+      };
+  }
+}
+
+base class TimeEntity private (
+  protected hours: Int,
+  protected minutes: UInt8,
+  protected seconds: UInt8,
+) {
+  //
+  fun getHours(): Int {
+    this.hours
+  }
+
+  fun getMinutes(): Int {
+    this.minutes.toInt()
+  }
+
+  fun getSeconds(): Int {
+    this.seconds.toInt()
+  }
+
+  fun addHours(hours: Int): this {
+    !this.hours = this.hours + hours;
+    this
+  }
+
+  fun addMinutes(minutes: Int): this {
+    if (minutes == 0) {
+      this
+    } else {
+      tminutes = this.minutes.toInt() + minutes;
+      aminutes = (Math.abs(tminutes) % 60);
+      rminutes = if (tminutes >= 0) {
+        aminutes
+      } else {
+        60 - aminutes
+      };
+      !this.minutes = UInt8::truncate(rminutes);
+      this.addHours(
+        if (tminutes >= 0) {
+          tminutes / 60
+        } else {
+          (tminutes / 60) - 1
+        },
+      );
+    };
+  }
+
+  fun addSeconds(second: Int): this {
+    if (second == 0) {
+      this
+    } else {
+      tsecond = this.seconds.toInt() + second;
+      asecond = (Math.abs(tsecond) % 60);
+      rsecond = if (tsecond >= 0) {
+        asecond
+      } else {
+        60 - asecond
+      };
+      !this.seconds = UInt8::truncate(rsecond);
+      this.addMinutes(
+        if (tsecond >= 0) {
+          tsecond / 60
+        } else {
+          (tsecond / 60) - 1
+        },
+      );
+    };
+  }
+}
+
```

Comment:
There's some inconsistency here where `add{Minutes|Seconds|Hours}` mutate their receiver, and `add{Days|Months|Years}` above leave their receiver unchanged and construct/return a new `Date`.

For what it's worth, I much prefer the immutable interface

Reply:
The class is not mutable, so the receiver is not mutated.